src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/InterpolatedStringBuilder.cs

```diff
@@ -0,0 +1,605 @@
+
+            bool leftAlign = false;
+            if (alignment < 0)
+            {
+                leftAlign = true;
+                alignment = -alignment;
+            }
+
+            int paddingNeeded = alignment - charsWritten;
+            if (paddingNeeded > 0)
+            {
+                EnsureCapacityForAdditionalChars(paddingNeeded);
+
+                if (leftAlign)
+                {
+                    _chars.Slice(_pos, paddingNeeded).Fill(' ');
+                }
+                else
+                {
+                    _chars.Slice(startingPos, charsWritten).CopyTo(_chars.Slice(startingPos + paddingNeeded));
+                    _chars.Slice(startingPos, paddingNeeded).Fill(' ');
+                }
+
+                _pos += paddingNeeded;
+            }
+        }
+
+        /// <summary>Ensures <see cref="_chars"/> has the capacity to store <paramref name="additionalChars"/> beyond <see cref="_pos"/>.</summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void EnsureCapacityForAdditionalChars(int additionalChars)
+        {
+            if (_chars.Length - _pos < additionalChars)
+            {
+                Grow(additionalChars);
+            }
+        }
+
+        /// <summary>Fallback for fast path in <see cref="AppendLiteral"/> when there's not enough space in the destination.</summary>
+        /// <param name="value">The string to write.</param>
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        private void GrowThenCopyString(string value)
+        {
+            Grow(value.Length);
+            value.CopyTo(_chars.Slice(_pos));
+            _pos += value.Length;
+        }
+
+        /// <summary>Fallback for <see cref="AppendFormatted(ReadOnlySpan{char})"/> for when not enough space exists in the current buffer.</summary>
+        /// <param name="value">The span to write.</param>
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        private void GrowThenCopySpan(ReadOnlySpan<char> value)
+        {
+            Grow(value.Length);
+            value.CopyTo(_chars.Slice(_pos));
+            _pos += value.Length;
+        }
+
+        /// <summary>Grows <see cref="_chars"/> to have the capacity to store <paramref name="additionalChars"/> beyond <see cref="_pos"/>.</summary>
+        [MethodImpl(MethodImplOptions.NoInlining)] // keep consumers as streamlined as possible
+        private void Grow(int additionalChars)
+        {
+            Debug.Assert(_pos > _chars.Length - additionalChars);
+
+            // We want the max of doubling our capacity and how much space we know we actually required. However,
+            // if doubling would exceed the max array length possible, stick to the number of characters required.
+            uint doubling = (uint)_chars.Length * 2;
+            uint required = (uint)_pos + (uint)additionalChars;
+            GrowCore(doubling > Array.MaxArrayLength || required > doubling ? required : doubling);
```

Comment:
FWIW string lengths are limited to a different max than normal arrays. See:

https://github.com/dotnet/runtime/blob/53d3bddc936e16b34fbe71cfd33964c83709600c/src/coreclr/vm/gchelpers.cpp#L846-L851

We don't have any accelerator for this like we have for `Array.MaxArrayLength`. Maybe it'd make sense to include an internal one here?

Why the different behavior between this and `Grow()`? Is there any risk that this could result in an O(n^2) complexity instead of amortized O(n)? I see you used `Clamp` below to work around a similar issue. (Though TBH we should consider capping at the same max string length, because there's no point to renting if we know _ToString_ is going to fail.)

Reply:
> Maybe it'd make sense to include an internal one here?

Added.

> Why the different behavior between this and Grow()?

Here, we know exactly how many characters are required for the operation in flight to be successful: anything less and it will fail, and the caller assumes if Grow returns successfully it will have allocated that much space, so while we'd like to double, at a minimum we need the requested number.  Grow() is called when we know we need more space but we don't know how much, so we'd like to double, but we're ok growing by a smaller amount as long as it's actually increasing the size.  That said, I can simplify what's there.

> Though TBH we should consider capping at the same max string length, because there's no point to renting if we know ToString is going to fail.

My concern there is if the caller actually needed string.MaxLength + 1 characters: we don't want to clamp it to string.MaxLength, in which case it might actually successfully allocate but then be less than the space needed by the caller, who is going to proceed to write into the space assuming it was sufficient.